drivers/sensors/paw3204.c

```diff
@@ -0,0 +1,122 @@
+/* Copyright 2021 Gompa (@Gompa)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "paw3204.h"
+#include "wait.h"
+#include "debug.h"
+#include "gpio.h"
+
+uint8_t datatogglestate;
+#define REG_PID1 0x00
+#define REG_PID2 0x01
+#define REG_STAT 0x02
+#define REG_X 0x03
+#define REG_Y 0x04
+
+#define REG_SETUP 0x06
+#define REG_IMGQUAL 0x07
+#define REG_IMGREC 0x0E
+#define REG_IMGTRASH 0x0D
+
+void PAW3204_init(void) {
+    setPinOutput(PAW3204_SCLK);    // setclockpin to output
+    setPinInputHigh(PAW3204_DATA); // set datapin input high
+
+    PAW3204_write_reg(REG_SETUP, 0x86); // reset sensor and set 1600cpi
+    wait_us(5);
+
+    PAW3204_read_reg(0x00); // read id
+    PAW3204_read_reg(0x01); // read id2
+    // PAW3204_write_reg(REG_SETUP,0x06);  // dont reset sensor and set cpi 1600
+    PAW3204_write_reg(REG_IMGTRASH, 0x32); // write image trashhold
+}
+
+uint8_t PAW3204_serial_read(void) {
+    setPinInput(PAW3204_DATA);
+    uint8_t byte = 0;
+
+    for (uint8_t i = 0; i < 8; ++i) {
+        writePinLow(PAW3204_SCLK);
+        wait_us(1);
+
+        byte = (byte << 1) | readPin(PAW3204_DATA);
+
+        writePinHigh(PAW3204_SCLK);
+        wait_us(1);
+    }
+
+    return byte;
+}
+
+void PAW3204_serial_write(uint8_t data) {
+    datatogglestate = readPin(PAW3204_DATA);
+    if (datatogglestate == 1) {
+        writePinLow(PAW3204_DATA);
+    } else {
+        writePinLow(PAW3204_DATA);
+    }
+    setPinOutput(PAW3204_DATA);
+
+    for (int8_t b = 7; b >= 0; b--) {
+        writePinLow(PAW3204_SCLK);
+        if (data & (1 << b))
+            writePinHigh(PAW3204_DATA);
+        else
+            writePinLow(PAW3204_DATA);
+        // wait_us(2);
+        writePinHigh(PAW3204_SCLK);
+    }
+
+    wait_us(4);
+}
+
+int8_t convert_twoscomp(uint8_t data) {
```

Comment:
is the value for the sensor only able to return -127 - 127? 

Eg, if it's capable of a higher value, then I would prefer that this is grabbed and handed over, and then let the pointing_device_driver.c code handle constraining the limit. 

This conforms better to the existing behavior, and makes #16371 much more "Drop in" for this sensor.

Reply:
according to the data sheet the report range is  –128 ~ +127. what would the best way of doing this be ?

Reply:
then, as is is fine.  It just means that it won't benefit from the linked PR, and that's fine.  my concern was if the sensor was capable of it, at all (some of the other ones are).